addons/purchase_mrp/report/mrp_report_bom_structure.py

```diff
@@ -25,6 +25,10 @@ def _format_route_info(self, rules, rules_delay, warehouse, product, bom, quanti
                 }
         return res
 
+    @api.model
+    def _is_resupply_rules(self, rules, bom):
+        return super()._is_resupply_rules(rules, bom) or any(rule for rule in rules if rule.action == 'buy')
```

Comment:
Maybe you can add `product` in the arguments so you can reuse `_is_buy_route`
```suggestion
    def _is_resupply_rules(self, rules, product, bom):
        return super()._is_resupply_rules(rules, bom) or self._is_buy_route(rules, product, bom)
```

Reply:
I don't know why I hesitated. You're right, changing it.

Reply:
EDIT: I knew there was a reason! We can't use `_is_buy_route()`, as this one is overriden in `mrp_subcontracting` to specifically *exclude* subcontracting from that. The reason being that in `_format_route_info()`, we don't want to format a subcontracting-specific route (e.g. `Dropship Subcontractor on Order`) as being a "Buy".
But in the context of `_find_special_rules()`, we're specifically looking for subcontracting-specific routes.

Reply:
Indeed, I didn't see the override :sweat_smile: 
In that case, I suggest you to just make the condition a little bit shorter :pinching_hand: 
```suggestion
    def _is_resupply_rules(self, rules, bom):
        return super()._is_resupply_rules(rules, bom) or any(rule == 'buy' for rule in rules)
```